pybind11/setup_helpers.py

```diff
@@ -306,24 +324,42 @@ class ParallelCompile(object):
     This takes several arguments that allow you to customize the compile
     function created:
 
-    envvar: Set an environment variable to control the compilation threads, like NPY_NUM_BUILD_JOBS
-    default: 0 will automatically multithread, or 1 will only multithread if the envvar is set.
-    max: The limit for automatic multithreading if non-zero
+    envvar:
+        Set an environment variable to control the compilation threads, like
+        NPY_NUM_BUILD_JOBS
+    default:
+        0 will automatically multithread, or 1 will only multithread if the
+        envvar is set.
+    max:
+        The limit for automatic multithreading if non-zero
+    needs_recompile:
+        A function of (obj, src) that returns True when recompile is needed.  No
+        effect in isolated mode; use ccache instead, see
+        https://github.com/matplotlib/matplotlib/issues/1507/
+
+    To use::
 
-    To use:
         ParallelCompile("NPY_NUM_BUILD_JOBS").install()
-    or:
+
+    or::
+
         with ParallelCompile("NPY_NUM_BUILD_JOBS"):
             setup(...)
+
+    By default, this assumes all files need to be recompiled. A smarter
+    function can be provided via needs_recompile.  If the output has not yet
+    been generated, the compile will always run, and this function is not
+    called.
     """
 
-    __slots__ = ("envvar", "default", "max", "old")
+    __slots__ = ("envvar", "default", "max", "_old", "needs_recompile")
 
-    def __init__(self, envvar=None, default=0, max=0):
+    def __init__(self, envvar=None, default=0, max=0, needs_recompile=no_recompile):
```

Comment:
Oh, interesting. I was just expecting `needs_recompile=None` as default.
But since this is nothing performance-critical, why not like this, yes.

Reply:
This provided slightly simpler typing, being just `Callable[[str, str],bool ` instead of `Optional[Callable[[str, str],bool]`, which is why I did it this way.